File: src/PhpProxyHunter/FileLockHelper.php
<?php

namespace PhpProxyHunter;

class FileLockHelper {
  public string $filePath;
  private $handle;
  private int $lockType;

  /**
   * Constructor.
   *
   * @param string $filePath Path to the lock file.
   */
  public function __construct(string $filePath) {
    $this->filePath = $filePath;
  }

  /**
   * Acquire a lock.
   *
   * @param int $lockType Use LOCK_EX for exclusive, LOCK_SH for shared.
   * @return bool True on success, false on failure.
   */
  public function lock(int $lockType = LOCK_EX): bool {
    $this->handle = $this->openLockHandle();
    if ($this->handle === false) {
      return false;
    }

    $this->lockType = $lockType;
    if (!flock($this->handle, $lockType)) {
      fclose($this->handle);
      $this->handle = null;
      return false;
    }

    return true;
  }

  /**
   * Unlock and close the file handle.
   *
   * @return void
   */
  public function unlock(): void {
    if (is_resource($this->handle)) {
      flock($this->handle, LOCK_UN);
      fclose($this->handle);
      $this->handle = null;
      // delete the lock file gracefully
      @unlink($this->filePath);
    }
  }

  /**
   * Release the lock.
   *
   * @return void
   */
  public function release(): void {
    $this->unlock();
  }

  /**
   * Check if the lock file currently exists.
   *
   * @return bool
   */
  public function isLocked(): bool {
    // Attempt an exclusive non-blocking lock. This will fail if any other
    // process holds either a shared or exclusive lock on the file, so it
    // conservatively indicates "locked by someone else".
    $handle = $this->openLockHandle();
    if ($handle === false) {
      // Can't open the file — fall back to file existence check
      return file_exists($this->filePath);
    }

    $isLocked = !flock($handle, LOCK_EX | LOCK_NB);
    if (!$isLocked) {
      // We acquired the exclusive lock; release it immediately
      flock($handle, LOCK_UN);
    }

    fclose($handle);
    return $isLocked;
  }

  /**
   * Open the lock file handle, ensuring the directory exists and is writable.
   * Returns a resource (handle) on success or false on failure.
   *
   * @return resource|false
   */
  private function openLockHandle() {
    $dir = dirname($this->filePath);

    if (!is_dir($dir)) {
      if (!@mkdir($dir, 0777, true) && !is_dir($dir)) {
        return false;
      }
    }

    if (!is_writable($dir)) {
      return false;
    }

    $handle = @fopen($this->filePath, 'c+');
    if ($handle === false) {
      return false;
    }

    return $handle;
  }

  /**
   * Automatically unlock on destruction.
   */
  public function __destruct() {
    $this->unlock();
  }
}

// Only run when executed directly from CLI, not when included or required
if (php_sapi_name() === 'cli' && realpath(__FILE__) === realpath($_SERVER['argv'][0] ?? '')) {
  $lockFile = __DIR__ . '/mylock.lock';
  $lock     = new FileLockHelper($lockFile);

  if ($lock->lock(LOCK_EX)) {
    echo "Lock acquired {$lock->filePath}. Starting work...\n";

    // Simulate work loop with periodic output
    for ($i = 1; $i <= 5; $i++) {
      echo "Working... step $i\n";
      if ($lock->isLocked()) {
        echo "The lock file is in use by another process.\n";
      } else {
        echo "The lock file is free.\n";
      }
      if ($lock->isLocked()) {
        echo "The lock file exists.\n";
      } else {
        echo "The lock file does not exist.\n";
      }
      sleep(1); // Simulate task delay
    }

    echo "Work done. Releasing lock...\n";
  // No need to manually call unlock() — __destruct will handle it
  } else {
    echo "Another instance is already running. Exiting.\n";
    exit(1);
  }
}

File: src/utils/file/FileLockHelper.py
"""
File Lock Helper

Provides robust file-based locking for process synchronization.
Similar to PHP FileLockHelper but using platform-specific locking mechanisms.
"""

import os
import sys
from pathlib import Path

# Platform-specific imports
if sys.platform == "win32":
    import msvcrt
else:
    import fcntl


class FileLockHelper:
    """File-based lock helper for process synchronization."""

    def __init__(self, file_path: str):
        """
        Constructor.

        Args:
            file_path (str): Path to the lock file.
        """
        self.file_path = file_path
        self.handle = None
        self.lock_type = None

    def lock(self, lock_type: str = "exclusive") -> bool:
        """
        Acquire a lock.

        Args:
            lock_type (str): Use 'exclusive' for exclusive, 'shared' for shared lock.

        Returns:
            bool: True on success, False on failure.
        """
        self.handle = self._open_lock_handle()
        if self.handle is None:
            return False

        self.lock_type = lock_type

        try:
            # Use non-blocking lock so callers don't hang waiting for the lock
            # to be released by another process. Return False immediately if
            # the lock can't be acquired.
            if sys.platform == "win32":
                # Windows: Use msvcrt for file locking
                try:
                    msvcrt.locking(self.handle.fileno(), msvcrt.LK_NBLCK, 1)
                    return True
                except OSError:
                    self.handle.close()
                    self.handle = None
                    return False
            else:
                # Unix-like: Use fcntl for file locking
                lock_flag = fcntl.LOCK_EX if lock_type == "exclusive" else fcntl.LOCK_SH
                fcntl.flock(self.handle, lock_flag | fcntl.LOCK_NB)
                return True
        except (IOError, OSError):
            # Lock failed, close the handle
            if self.handle:
                self.handle.close()
                self.handle = None
            return False

    def unlock(self) -> None:
        """Unlock and close the file handle."""
        if self.handle is not None:
            try:
                if sys.platform == "win32":
                    # Windows: Unlock
                    try:
                        msvcrt.locking(self.handle.fileno(), msvcrt.LK_UNLCK, 1)
                    except OSError:
                        pass
                else:
                    # Unix-like: Unlock
                    fcntl.flock(self.handle, fcntl.LOCK_UN)
                self.handle.close()
            except (IOError, OSError):
                pass
            finally:
                self.handle = None

            # Delete the lock file gracefully
            try:
                if os.path.exists(self.file_path):
                    os.unlink(self.file_path)
            except (IOError, OSError):
                pass

    def release(self) -> None:
        """Release the lock (alias for unlock)."""
        self.unlock()

    def is_locked(self) -> bool:
        """
        Check if the lock file is currently locked by another process.

        Returns:
            bool: True if locked by another process, False otherwise.
        """
        # Attempt an exclusive non-blocking lock. This will fail if any other
        # process holds either a shared or exclusive lock on the file.
        handle = self._open_lock_handle()
        if handle is None:
            # Can't open the file — fall back to file existence check
            return os.path.exists(self.file_path)

        try:
            if sys.platform == "win32":
                try:
                    msvcrt.locking(handle.fileno(), msvcrt.LK_NBLCK, 1)
                    # We acquired the lock; release it immediately
                    msvcrt.locking(handle.fileno(), msvcrt.LK_UNLCK, 1)
                    handle.close()
                    return False
                except OSError:
                    # Lock is held by another process
                    handle.close()
                    return True
            else:
                fcntl.flock(handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
                # We acquired the exclusive lock; release it immediately
                fcntl.flock(handle, fcntl.LOCK_UN)
                handle.close()
                return False
        except (IOError, OSError):
            # Lock is held by another process
            try:
                handle.close()
            except (IOError, OSError):
                pass
            return True

    def _open_lock_handle(self):
        """
        Open the lock file handle, ensuring the directory exists and is writable.
        Returns a file handle on success or None on failure.

        Returns:
            file object or None
        """
        lock_dir = os.path.dirname(self.file_path)

        if lock_dir and not os.path.isdir(lock_dir):
            try:
                os.makedirs(lock_dir, mode=0o777, exist_ok=True)
            except (IOError, OSError):
                return None

        if lock_dir and not os.access(lock_dir, os.W_OK):
            return None

        try:
            handle = open(self.file_path, "a+")
            return handle
        except (IOError, OSError):
            return None

    def __del__(self):
        """Automatically unlock on destruction."""
        self.unlock()


# Only run when executed directly from CLI
if __name__ == "__main__":
    lock_file = os.path.join(os.path.dirname(__file__), "mylock.lock")
    lock = FileLockHelper(lock_file)

    if lock.lock("exclusive"):
        print(f"Lock acquired {lock.file_path}. Starting work...")

        # Simulate work loop with periodic output
        import time

        for i in range(1, 6):
            print(f"Working... step {i}")
            if lock.is_locked():
                print("The lock file is in use by another process.")
            else:
                print("The lock file is free.")
            print(f"Lock file exists: {os.path.exists(lock.file_path)}")
            time.sleep(1)

        lock.unlock()
        print("Lock released")
    else:
        print(f"Could not acquire lock on {lock_file}")
